### PR TITLE
Bump astroid to 3.0.0, update changelog

### LINKED ISSUE #679: Remove monkey-patching of methods onto classes
In [this comment](https://github.com/PyCQA/astroid/issues/170#issuecomment-163118573), @PCManticore stated that there was a plan to remove monkey-patching of methods onto classes. inference.py and protocols.py seem to suggest that this hasn't happened.

Is this still a design goal?

It looks to me like this would be a fairly mechanical change. I am assuming that since it hasn't happened yet, I am missing something. Is it simply a "haven't had time" issue?

### LINKED ISSUE #1072: Remove deprecated ``astroid.node_classes`` and ``astroid.scoped_nodes``
Those two files have been deprecated in 2.7.0 and need to be removed in 3.0.0.

### LINKED ISSUE #1780: Astroid is leaking memory
Astroid is leaking memory when is run multiple times over same file using API

### Steps to reproduce

1. Pick larger file, e.g. file from pylint project: pylint/checkers/variables.py
2. Run following script:
```python
from pylint.lint import Run
from pylint.lint import pylinter
while True:
    Run(["pylint/checkers/variables.py"], exit=False)
    pylinter.MANAGER.clear_cache()
```

### Current behavior

After ~12 iteration memory goes over 1 GB and keeps growing
image

### Expected behavior

Allocated memory should not rise indefinitelly

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output

2.13.0-dev0

### LINKED ISSUE #1828: Wrong type inferred - inference type cache is missing context
Moving discussion started in https://github.com/PyCQA/pylint/issues/7464 to astroid repo.

### Steps to reproduce

```
import astroid

result1, result2 = astroid.extract_node("""
class Base:
    def return_type(self):
        return type(self)()

class A(Base):
    def method(self):
        return self.return_type()

class B(Base):
    def method(self):
        return self.return_type()

A().method()  #@
B().method()  #@
""")

print(next(result1.infer()))
print(next(result2.infer()))
```

### Current behavior
Instance of .A
Instance of .A

### Expected behavior
Instance of .A
Instance of .B

### Workaround
Modify inference_tip.py to disable cache in _inference_tip_cached function. 

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output
2.13.0-dev0

### LINKED ISSUE #2161: Make ``import astroid`` faster (again)
### Steps to reproduce

```
python3 -X importtime -c 'import astroid'
scalene --html pylint/__main__.py --- pylint/constants.py
```
import_astroid_too_slow

```
import time: self [us] | cumulative | imported package
import time:       157 |        157 |   _io
import time:        47 |         47 |   marshal
import time:       351 |        351 |   posix
import time:       965 |       1518 | _frozen_importlib_external
import time:        96 |         96 |   time
import time:       167 |        262 | zipimport
import time:       122 |        122 |     _codecs
import time:       726 |        848 |   codecs
import time:       497 |        497 |   encodings.aliases
import time:      1443 |       2787 | encodings
import time:       156 |        156 | encodings.utf_8
import time:        66 |         66 | _signal
import time:        36 |         36 |     _abc
import time:       252 |        288 |   abc
import time:       548 |        835 | io
import time:        41 |         41 |       _stat
import time:       146 |        186 |     stat
import time:       738 |        738 |     _collections_abc
import time:       106 |        106 |       genericpath
import time:       255 |        360 |     posixpath
import time:      3296 |       4579 |   os
import time:       127 |        127 |   _sitebuiltins
import time:       273 |        273 |     apport_python_hook
import time:       132 |        404 |   sitecustomize
import time:      1287 |       6395 | site
import time:       127 |        127 |       itertools
import time:       110 |        110 |       keyword
import time:        51 |         51 |         _operator
import time:       219 |        270 |       operator
import time:       133 |        133 |       reprlib
import time:        46 |         46 |       _collections
import time:      1002 |       1684 |     collections
import time:       199 |        199 |     types
import time:        41 |         41 |     _functools
import time:       509 |       2432 |   functools
import time:       529 |        529 |       enum
import time:        67 |         67 |         _sre
import time:       203 |        203 |           sre_constants
import time:       300 |        502 |         sre_parse
import time:      1262 |       1830 |       sre_compile
import time:        54 |         54 |       _locale
import time:       139 |        139 |       copyreg
import time:       569 |       3119 |     re
import time:       135 |        135 |     token
import time:       848 |       4101 |   tokenize
import time:       190 |        190 |     warnings
import time:       269 |        459 |   importlib
import time:       124 |        124 |       __future__
import time:       117 |        117 |       collections.abc
import time:       434 |        434 |         contextlib
import time:      1557 |       1990 |       typing
import time:       351 |        351 |         astroid.util
import time:      1682 |       1682 |         astroid.typing
import time:       892 |       2924 |       astroid.exceptions
import time:       593 |        593 |                 _weakrefset
import time:       912 |       1504 |               weakref
import time:        91 |         91 |                   org
import time:        40 |        131 |                 org.python
import time:        38 |        168 |               org.python.core
import time:       253 |       1925 |             copy
import time:       154 |        154 |                 _ast
import time:      1483 |       1636 |               ast
import time:       143 |        143 |                   _opcode
import time:       285 |        428 |                 opcode
import time:       432 |        859 |               dis
import time:       240 |        240 |               importlib.machinery
import time:       134 |        134 |               linecache
import time:      2600 |       5468 |             inspect
import time:      1872 |       9264 |           dataclasses
import time:       408 |       9672 |         pprint
import time:       281 |        281 |         astroid.context
import time:       670 |        670 |             importlib._abc
import time:       217 |        886 |           importlib.util
import time:       131 |        131 |               fnmatch
import time:       120 |        120 |                 _winapi
import time:        57 |         57 |                 nt
import time:        47 |         47 |                 nt
import time:        46 |         46 |                 nt
import time:        93 |         93 |                 nt
import time:       439 |        800 |               ntpath
import time:        54 |         54 |               errno
import time:       102 |        102 |                 urllib
import time:      1342 |       1444 |               urllib.parse
import time:      1043 |       3470 |             pathlib
import time:       286 |       3756 |           astroid.const
import time:       123 |        123 |             astroid.interpreter
import time:       183 |        306 |           astroid.interpreter._import
import time:       384 |        384 |                 traceback
import time:       225 |        225 |                   _string
import time:      3610 |       3834 |                 string
import time:       910 |        910 |                 threading
import time:       116 |        116 |                 atexit
import time:     10004 |      15246 |               logging
import time:       678 |        678 |               sysconfig
import time:       234 |        234 |               astroid.interpreter._import.util
import time:       587 |        587 |               _sysconfigdata__x86_64-linux-gnu
import time:      2256 |      19000 |             astroid.modutils
import time:       985 |      19985 |           astroid.interpreter._import.spec
import time:       335 |        335 |           astroid.transforms
import time:       502 |      25767 |         astroid.manager
import time:       898 |        898 |         astroid.nodes.as_string
import time:       134 |        134 |         astroid.nodes.const
import time:       240 |        240 |         astroid.nodes.utils
import time:       876 |      37865 |       astroid.nodes.node_ng
import time:      7112 |      50129 |     astroid.nodes._base_nodes
import time:       506 |        506 |       astroid.decorators
import time:       822 |        822 |         astroid.interpreter.objectmodel
import time:       728 |       1550 |       astroid.bases
import time:      3685 |       5740 |     astroid.nodes.node_classes
import time:       121 |        121 |         astroid.filter_statements
import time:       112 |        112 |         astroid.nodes.scoped_nodes.utils
import time:       291 |        522 |       astroid.nodes.scoped_nodes.mixin
import time:       103 |        103 |         astroid.interpreter.dunder_lookup
import time:       657 |        759 |       astroid.nodes.scoped_nodes.scoped_nodes
import time:       283 |       1564 |     astroid.nodes.scoped_nodes
import time:       527 |      57958 |   astroid.nodes
import time:       803 |        803 |       typing_extensions
import time:       230 |       1032 |     astroid.constraint
import time:    119122 |     119122 |       astroid.raw_building
import time:       247 |     119368 |     astroid.helpers
import time:       436 |        436 |     astroid.objects
import time:       237 |        237 |       astroid.arguments
import time:       306 |        543 |     astroid.protocols
import time:       600 |     121978 |   astroid.inference
import time:        93 |         93 |   astroid.__pkginfo__
import time:        84 |         84 |   astroid.astroid_manager
import time:        87 |         87 |     astroid.brain
import time:       189 |        275 |   astroid.brain.helpers
import time:       967 |        967 |     textwrap
import time:       370 |        370 |       astroid._ast
import time:       511 |        880 |     astroid.rebuilder
import time:       275 |       2120 |   astroid.builder
import time:       123 |        123 |   astroid.inference_tip
import time:        99 |         99 |   astroid.brain.brain_numpy_utils
import time:        49 |         49 |     math
import time:        27 |         27 |       _bisect
import time:       141 |        167 |     bisect
import time:        28 |         28 |     _random
import time:        26 |         26 |     _sha512
import time:       418 |        687 |   random
import time:     12916 |     203319 | astroid
```

### Current behavior

Currently astroid import time account for 25+% of the parse time for single file analysis in pylint (common with parallelized pre-commit hook). This is 50+% with ``--disable=all``. This is a follow up to #1320 where the import time was already decreased by 60%. We can see that importing ``astroid.raw_building`` now takes 60% of the remaining import time, ``astroid.nodes._base_nodes`` (3.5%) and ``astroid.nodes.node_classes`` (1,7%) are the next biggest offenders that we control. 

### Expected behavior

Faster astroid import to make pylint more reactive.

### LINKED ISSUE #2201: Implement new nodes for PEP 695: Type Parameter Syntax
There's a new syntax in python 3.12, we need to handle it before claiming we support 3.12, see https://docs.python.org/3.12/whatsnew/3.12.html#pep-695-type-parameter-syntax

### LINKED ISSUE #2213: `.arguments` property ignores keyword-only args, *args, and **kwargs
```python
>>> from astroid import extract_node
>>> node = extract_node("""def a(*args, b=None, c=None, **kwargs): ...""")
>>> node.args.arguments
[]
```

Expected to find all the arguments from the function signature.

The wanted data can be found here:

```python
>>> node.args.vararg
'args'
>>> node.args.kwarg
'kwargs'
>>> node.args.kwonlyargs
[<AssignName.b l.1 at 0x1048189b0>, <AssignName.c l.1 at 0x104818830>]
```

Discussed at https://github.com/pylint-dev/pylint/pull/7577#discussion_r989000829.

Notice that positional-only args are found for some reason 🤷

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
* Add support for Python 3.12, including PEP 695 type parameter syntax.

  Closes #2201

* Remove support for Python 3.7.

  Refs #2137

* Use the global inference cache when inferring, even without an explicit
  ``InferenceContext``. This is a significant performance improvement given how
  often methods default to ``None`` for the context argument. (Linting ``astroid``
  itself now takes ~5% less time on Python 3.12; other projects requiring more
  complex inference calculations will see greater speedups.)

  Refs #529

* Following a deprecation period starting in astroid 2.7.0, the ``astroid.node_classes``
  and ``astroid.scoped_nodes`` modules have been removed in favor of ``astroid.nodes.node_classes``
  and ``astroid.nodes.scoped_nodes``.

  Closes #1072

* Following a deprecation period starting in astroid 2.12.0, the ``astroid.mixins`` module
  has been removed in favor of ``astroid.nodes._base_nodes`` (private).

  Refs #1633

* Return all existing arguments when calling ``Arguments.arguments()``. This also means ``find_argname`` will now
  use the whole list of arguments for its search.

  Closes #2213

* Exclude class attributes from the ``__members__`` container of an ``Enum`` class when they are
  ``nodes.AnnAssign`` nodes with no assigned value.

  Refs pylint-dev/pylint#7402

* Remove ``@cached`` and ``@cachedproperty`` decorator (just use ``@cached_property`` from the stdlib).

  Closes #1780
  Refs #2140

* Remove the ``inference`` module. Node inference methods are now in the module
  defining the node, rather than being associated to the node afterward.

  Closes #679

* Move ``LookupMixIn`` to ``astroid.nodes._base_nodes`` and make it private.

* Remove the shims for ``OperationError``, ``BinaryOperationError``, and ``UnaryOperationError``
  in ``exceptions``. They were moved to ``util`` in astroid 1.5.0.

* Move ``safe_infer()`` from ``helpers`` to ``util``. This avoids some circular imports.

* Reduce file system access in ``ast_from_file()``.

* Reduce time to ``import astroid`` by delaying ``astroid_bootstrapping()`` until
  the first instantiation of ``AstroidBuilder``.

  Closes #2161

* Make ``igetattr()`` idempotent. This addresses some reports of varying results
  when running pylint with ``--jobs``.

  Closes pylint-dev/pylint#4356
  Refs #7

* Fix incorrect cache keys for inference results, thereby correctly inferring types
  for calls instantiating types dynamically.

  Closes #1828
  Closes pylint-dev/pylint#7464
  Closes pylint-dev/pylint#8074

* Fix interrupted ``InferenceContext`` call chains, thereby addressing performance
  problems when linting ``sqlalchemy``.

  Closes pylint-dev/pylint#8150

* ``nodes.FunctionDef`` no longer inherits from ``nodes.Lambda``.
  This is a breaking change but considered a bug fix as the nodes did not share the same
  API and were not interchangeable.

  We have tried to minimize the amount of breaking changes caused by this change
  but some are unavoidable.

* ``infer_call_result`` now shares the same interface across all implementations. Namely:
  ```python
  def infer_call_result(
                self,
                caller: SuccessfulInferenceResult | None,
                context: InferenceContext | None = None,
            ) -> Iterator[InferenceResult]:
  ```

  This is a breaking change for ``nodes.FunctionDef`` where previously ``caller`` had a default of
  ``None``. Passing ``None`` again will not create a behaviour change.
  The breaking change allows us to better type and re-use the method within ``astroid``.

* Improved signature of the ``__init__`` and ``__postinit__`` methods of most nodes.
  This includes making ``lineno``, ``col_offset``, ``end_lineno``, ``end_col_offset`` and ``parent``
  required arguments for ``nodes.NodeNG`` and its subclasses.
  For most other nodes, arguments of their ``__postinit__`` methods have been made required to better
  represent how they would normally be constructed by the standard library ``ast`` module.
  The following nodes were changed or updated:
  - ``nodes.AnnAssign``
  - ``nodes.Arguments``
  - ``nodes.Assign``
  - ``nodes.AssignAttr``
  - ``nodes.AssignName``
  - ``nodes.Attribute``
  - ``nodes.AugAssign``
  - ``nodes.Await``
  - ``nodes.BaseContainer``
  - ``nodes.BinOp``
  - ``nodes.Call``
  - ``nodes.ClassDef``
  - ``nodes.Compare``
  - ``nodes.Comprehension``
  - ``nodes.Decorators``
  - ``nodes.Delete``
  - ``nodes.DelAttr``
  - ``nodes.DelName``
  - ``nodes.Dict``
  - ``nodes.DictComp``
  - ``nodes.ExceptHandler``
  - ``nodes.Expr``
  - ``nodes.For``
  - ``nodes.FunctionDef``
  - ``nodes.GeneratorExp``
  - ``nodes.If``
  - ``nodes.IfExp``
  - ``nodes.Keyword``
  - ``nodes.Lambda``
  - ``nodes.ListComp``
  - ``nodes.Module``
  - ``nodes.Name``
  - ``nodes.NodeNG``
  - ``nodes.Raise``
  - ``nodes.Return``
  - ``nodes.SetComp``
  - ``nodes.Slice``
  - ``nodes.Starred``
  - ``objects.Super``, we also added the ``call`` parameter to its ``__init__`` method.
  - ``nodes.Subscript``
  - ``nodes.UnaryOp``
  - ``nodes.While``
  - ``nodes.Yield``

  These changes involve breaking changes to their API but should be considered bug fixes. We
  now make arguments required when they are instead of always providing defaults.

* ``nodes.If.self.is_orelse`` has been removed as it was never set correctly and therefore
  provided a false value.

* Remove dependency on ``wrapt``.

* Remove dependency on ``lazy_object_proxy``. This includes the removal
  of the associated ``lazy_import``, ``lazy_descriptor`` and ``proxy_alias`` utility functions.

* ``CallSite._unpack_args`` and ``CallSite._unpack_keywords`` now use ``safe_infer()`` for
  better inference and fewer false positives.

  Closes pylint-dev/pylint#8544

* Add ``attr.Factory`` to the recognized class attributes for classes decorated with ``attrs``.

  Closes pylint-dev/pylint#4341

* ``infer_property()`` now observes the same property-specific workaround as ``infer_functiondef``.

  Refs #1490

* Remove unused and / or deprecated constants:
  - ``astroid.bases.BOOL_SPECIAL_METHOD``
  - ``astroid.bases.BUILTINS``
  - ``astroid.const.BUILTINS``
  - ``astroid.const.PY38_PLUS``
  - ``astroid.const.Load``
  - ``astroid.const.Store``
  - ``astroid.const.Del``

  Refs #2141

* ``frame()`` raises ``ParentMissingError`` and ``statement()`` raises ``StatementMissing`` for
  missing parents regardless of the value of the ``future`` argument (which gave this behavior
  already).

  The ``future`` argument to each method is deprecated and will be removed in astroid 4.0.

  Refs #1217

* Remove deprecated ``Ellipsis``, ``ExtSlice``, ``Index`` nodes.

  Refs #2152

* Remove deprecated ``is_sys_guard`` and ``is_typing_guard`` methods.

  Refs #2153

* Remove deprecated ``doc`` attribute for ``Module``, ``ClassDef``, and ``FunctionDef``.
  Use the ``doc_node`` attribute instead.

  Refs #2154

* Add new ``nodes.Try`` to better match Python AST. Replaces the ``TryExcept``
  and ``TryFinally`` nodes which have been removed.

* Publicize ``NodeNG.repr_name()`` to facilitate finding a node's nice name.

  Refs pylint-dev/pylint#8598

* Fix false positives for ``no-member`` and ``invalid-name`` when using the ``_name_``, ``_value_`` and ``_ignore_`` sunders in Enums.

  Closes pylint-dev/pylint#9015

### CODE REVIEW

**Comment by jacobtylerwalls** [comment]
We should consider reverting #1873 as part of this. It was a good idea at the time, but we did a good job with the changelog fragments. I don't know what other content we would put in the upgrade guide.

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
Sorry, I had to put the date in. If it's ready today then.. I know the date 😄

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
Yeah, let's release a last 2.x maintenance version of astroid.

**Review by jacobtylerwalls — DISMISSED** [review]
ready for a rebase

**Review by DanielNoord — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
This will need a merge commit because I also merged the maintenance branch at the same time (squash or rebase won't do).

File: ChangeLog [hunk 1]
```diff
@@ -2,9 +2,22 @@
...
+What's New in astroid 3.0.1?
+============================
+Release date: TBA
+
+
+
 What's New in astroid 3.0.0?
 =============================
-Release date: TBA
+Release date: 2023-09-25
```
Comment (jacobtylerwalls) [inline ChangeLog:20]: off by one, but oh well :-)

Comment (Pierre-Sassoulas) [reply ChangeLog:20]: I also blundered the maintenance branch creation, let me fix that.

**Pull request merged**